goon.go

```diff
@@ -281,13 +262,26 @@ func (g *Goon) PutMulti(src interface{}) ([]*datastore.Key, error) {
 					mk := memkey(rkeys[i])
 					delete(g.toDelete, mk)
 					g.toSet[mk] = vi
+					g.toDeleteMC[mk] = true
 				} else {
 					g.putMemory(vi)
 				}
 			}
 		}(i)
 	}
 	wg.Wait()
+
+	// Memcache needs to be updated after the datastore to prevent a common race condition,
+	// where a concurrent request will fetch the not-yet-updated data from the datastore
+	// and populate memcache with it.
+	if !g.inTransaction {
+		var memkeys []string
+		for _, key := range keys {
+			memkeys = append(memkeys, memkey(key))
```

Comment:
Could you surround this `append` call with `if !key.Incomplete()` as it was previously? The `keys` slice can still contain incomplete keys when some entity put failed.